core/templates/dev/head/pages/exploration-editor-page/improvements-tab/services/improvement-modal.service.ts

```diff
@@ -48,6 +48,177 @@ angular.module('oppia').factory('ImprovementModalService', [
+
+              // This is the index of the first action in the range of actions
+              // displayed. When this index reaches 0, we've displayed all the
+              // learner actions.
+              let indexOfFirstDisplayedAction = playthrough.actions.length - 1;
+
+              /**
+               * Identifies whether all learner actions have been displayed and
+               * no further expansion is possible.
+               * @returns {boolean}
+               */
+              $scope.canExpandActions = function() {
+                return indexOfFirstDisplayedAction !== 0;
+              };
+
+              /**
+               * Iterates through the learner actions and expands the set of
+               * actions to be displayed. If all learner actions have been
+               * handled, no more expansion is possible.
+               */
+              let expandActions = function() {
+                let i;
+                let previousStateName = null;
+                if (indexOfFirstDisplayedAction ===
+                    playthrough.actions.length) {
+                  previousStateName = (
+                    playthrough.actions[
+                      0].actionCustomizationArgs.state_name.value);
+                } else {
+                  let action = playthrough.actions[indexOfFirstDisplayedAction];
+                  previousStateName =
+                    action.actionCustomizationArgs.state_name.value;
+                }
+                for (i = indexOfFirstDisplayedAction; i >= 0; i--) {
+                  let action = playthrough.actions[i];
+                  let currentStateName =
+                      action.actionCustomizationArgs.state_name.value;
+
+                  if (currentStateName !== previousStateName) {
+                    // When the latest state name and the state name of the
+                    // current learner action differ, they are part of
+                    // different contexts. So, we identify whether the size
+                    // of the block has crossed the threshold number of actions
+                    // per block. If it has, we don't add more learner actions
+                    // to the block.
+                    if (indexOfFirstDisplayedAction - i >= (
+                      MAX_UNRELATED_ACTIONS_PER_BLOCK)) {
+                      indexOfFirstDisplayedAction = i + 1;
+                      break;
+                    }
+                  }
+                  previousStateName = currentStateName
+                }
+                // This is the case when all learner actions have been iterated
+                // over, and no more expansion is possible.
+                if (i === -1) {
+                  indexOfFirstDisplayedAction = 0;
+                }
+              };
+
+              expandActions();
+
+              // Index to know where to start highlighting actions.
+              let indexOfFirstDisplayedBlock = indexOfFirstDisplayedAction;
```

Comment:
No idea what the conceptual difference between these two variables is -- please explain. Do you even need it? Can you compute the block index on the fly?

Reply:
@seanlip We can't compute this on the fly. This is the first time we expand actions i.e. the first block we display. So, this variable is stored then and not computable at all points.

Reply:
So to make it more clear, I've renamed the first variable to indexOfActionInFirstDisplayedBlock.